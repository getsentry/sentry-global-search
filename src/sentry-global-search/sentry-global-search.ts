import { MultipleQueriesQuery } from '@algolia/client-search';
import algoliasearch, { SearchClient } from 'algoliasearch/lite';

import { Config, SearchHit, Hit, Result, Site } from './lib/types';
import { sites, defaultQueryParams } from './lib/config';

const errorType = `SentryGlobalSearchError`;

type QueryArgs = {
  path?: string;
  platforms?: string[];
  searchAllIndexes?: boolean;
};

type OptionalFilters = Array<string | string[]>;

type Optional<T, K extends keyof T> = Pick<Partial<T>, K> & Omit<T, K>;
type ConstructorConfig = string | Optional<Config, 'indexes'>;

class SentryGlobalSearch {
  configs: Config[];
  client: SearchClient;

  constructor(configs: ConstructorConfig[] = []) {
    // Complain if no configuration has been provided
    if (configs.length === 0 || !Array.isArray(configs)) {
      throw new Error(
        `${errorType}: SentryGlobalSearch must be initialized with an array that includes one or more of: ${sites
          .map(x => x.site)
          .join(', ')}.`
      );
    }

    // Validate configuration
    this.configs = configs.map(x => {
      const config = typeof x === 'string' ? { site: x as Site } : x;
      const defaults = sites.find(x => x.site === config.site);

      if (!!defaults) return { ...defaults, ...config };

      throw new Error(
        `${errorType}: unknown site "${config.site}" in config.include`
      );
    });

    // Create an Algolia client to work with
    this.client = algoliasearch(
      'OOK48W9UCL',
      '2d64ec1106519cbc672d863b0d200782'
    );
    this.query = this.query.bind(this);
  }

  async query(query: string, args: QueryArgs = {}) {
    if (!query) return [];

    // Strip out all but Basic Latin, to minimize impact from bot search that
    // uses random characters. We don't have localized docs so there's no point
    // in searching non-latin characters.
    const sanitizedQuery = query.replace(/[^\u0020-\u007f]/gi, '');

    const { client, configs } = this;

    const searchAllIndexes = args.searchAllIndexes || false;
    const configsToSearch = searchAllIndexes ? configs : [configs[0]];

    // Create a list of Algolia query objects from our configs
    const queries = configsToSearch.reduce<MultipleQueriesQuery[]>(
      (queries, config) => {
        const optionalFilters: OptionalFilters = [];

        if (config.pathBias && args.path) {
          optionalFilters.push(`pathSegments:${args.path}`);
        }

        if (
          config.platformBias &&
          args.platforms &&
          args.platforms.length > 0
        ) {
          optionalFilters.push(args.platforms.map(x => `platforms:${x}`));
        }

        if (config.legacyBias) {
          optionalFilters.push(`legacy:0`);
        }

        const newQueries = config.indexes.map(({ indexName }) => {
          const obj: MultipleQueriesQuery = {
            indexName,
            query: sanitizedQuery,
            params: {
              ...defaultQueryParams,
              ...(optionalFilters.length && { optionalFilters }),
            },
          };
          return obj;
        });
        return [...queries, ...newQueries];
      },
      []
    );

    // Get the search results
    const { results: algoliaResults } = await client.search<SearchHit>(queries);

    // Reduce and normalize the Algolia results
    const results = configsToSearch.map(config => {
      // If a site has more than one index, reduce them to one array.
      const hits = config.indexes.reduce<Hit[]>((hits, index) => {
        const algoliaResult = algoliaResults.find(
          result => result.index === index.indexName
        );

        // if no result return early
        if (!algoliaResult) {
          return [...hits];
        }

        // Normalize the results into a consistent format
        return [...hits, ...algoliaResult.hits.map(index.transformer)];
      }, []);

      return {
        site: config.site,
        name: config.name,
        hits,
      } as Result;
    });

    return results;
  }
}

export default SentryGlobalSearch;
