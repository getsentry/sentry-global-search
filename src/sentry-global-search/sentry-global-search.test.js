import { SentryGlobalSearch } from './sentry-global-search';
import algoliasearch from 'algoliasearch/lite';

const config = ['docs', 'develop', 'blog', 'help-center'];

describe('Search', () => {
  const client = algoliasearch();

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('requires configuration', () => {
    expect(() => {
      new SentryGlobalSearch();
    }).toThrowErrorMatchingSnapshot();
  });

  test('requires configuration as array', () => {
    expect(() => {
      const search = new SentryGlobalSearch(
        'docs',
        'develop',
        'blog',
        'help-center'
      );
    }).toThrowErrorMatchingSnapshot();
  });

  test('rejects unknown sites', () => {
    expect(() => {
      const search = new SentryGlobalSearch(['foobar']);
    }).toThrowErrorMatchingSnapshot();
  });

  test('returns nothing for an empty query', async () => {
    const search = new SentryGlobalSearch(config);
    const results = await search.query();
    expect(results).toEqual([]);
  });

  test('queries', async () => {
    const search = new SentryGlobalSearch(config);
    const results = await search.query('react');
    expect(results).toMatchSnapshot();
  });

  test('queries multiple indexes', async () => {
    const search = new SentryGlobalSearch(config);
    const results = await search.query('react', {
      searchAllIndexes: true,
    });
    expect(results).toMatchSnapshot();
  });

  test('queries with platform priority', async () => {
    const search = new SentryGlobalSearch(config);
    const results = await search.query('react', {
      platforms: ['sentry.javascript.react'],
    });
    expect(client.search.mock.calls).toMatchSnapshot();
  });

  test('does not query platforms when none are provided', async () => {
    const search = new SentryGlobalSearch(config);
    await search.query('react', {
      platforms: [],
    });
    expect(client.search.mock.calls).toMatchSnapshot();
  });

  test('queries with path as priority', async () => {
    const search = new SentryGlobalSearch(config);
    const results = await search.query('react', {
      path: ['/foo/bar/'],
    });
    expect(client.search.mock.calls).toMatchSnapshot();
  });

  test('prioritizes path over platform', async () => {
    const search = new SentryGlobalSearch(config);
    const results = await search.query('react', {
      path: ['/foo/bar/'],
      platforms: ['sentry.javascript.react'],
    });
    expect(client.search.mock.calls).toMatchSnapshot();
  });

  test('can bias individual sites', async () => {
    const biasedConfig = [...config];
    biasedConfig.push({
      site: biasedConfig.pop(),
      pathBias: true,
      platformBias: false,
      legacyBias: false,
    });
    const search = new SentryGlobalSearch(biasedConfig);
    const results = await search.query('react', {
      path: ['/foo/bar/'],
      platforms: ['sentry.javascript.react'],
      legacy: true,
    });
    expect(client.search.mock.calls).toMatchSnapshot();
  });
});
