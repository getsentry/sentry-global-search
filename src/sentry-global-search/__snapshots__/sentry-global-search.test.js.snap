// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search calls search with clickAnalytics 1`] = `
Array [
  Array [
    Array [
      Object {
        "indexName": "blog",
        "params": Object {
          "clickAnalytics": true,
          "highlightPostTag": "</mark>",
          "highlightPreTag": "<mark>",
          "optionalFilters": Array [
            Array [
              "platforms:sentry.javascript.react",
            ],
            "legacy:0",
          ],
          "snippetEllipsisText": "…",
        },
        "query": "react",
      },
    ],
  ],
]
`;

exports[`Search calls transformer with response 1`] = `
Object {
  "exhaustiveNbHits": true,
  "hits": Array [
    Object {
      "_highlightResult": Object {
        "excerpt": Object {
          "fullyHighlighted": false,
          "matchLevel": "full",
          "matchedWords": Array [
            "react",
          ],
          "value": "Mocked excerpt",
        },
        "fields": Object {
          "slug": Object {
            "fullyHighlighted": false,
            "matchLevel": "full",
            "matchedWords": Array [
              "react",
            ],
            "value": "/platforms/javascript/<mark>react</mark>/",
          },
        },
        "section": Object {
          "matchLevel": "none",
          "matchedWords": Array [],
          "value": "Mocked section",
        },
        "title": Object {
          "fullyHighlighted": true,
          "matchLevel": "full",
          "matchedWords": Array [
            "react",
          ],
          "value": "<mark>React</mark>",
        },
      },
      "_snippetResult": Object {
        "body_safe": Object {
          "matchLevel": "full",
          "value": "Mocked body_safe",
        },
        "content": Object {
          "matchLevel": "full",
          "value": "MockedContent",
        },
        "excerpt": Object {
          "matchLevel": "full",
          "value": "Mocked excerpt",
        },
        "text": Object {
          "matchLevel": "full",
          "value": "Mocked text",
        },
      },
      "draft": false,
      "excerpt": "Mocked excerpt",
      "fields": Object {
        "slug": "/platforms/javascript/react/",
      },
      "objectID": "cd5528ad-935b-5a33-b25e-4f945638551b",
      "section": Object {
        "full_path": "Mocked full_path",
      },
      "title": "Mocked hit",
      "url": "/platforms/javascript/react/",
    },
    Object {
      "_highlightResult": Object {
        "excerpt": Object {
          "fullyHighlighted": false,
          "matchLevel": "full",
          "matchedWords": Array [
            "react",
          ],
          "value": "Mocked excerpt",
        },
        "fields": Object {
          "slug": Object {
            "fullyHighlighted": false,
            "matchLevel": "full",
            "matchedWords": Array [
              "react",
            ],
            "value": "/platforms/javascript/<mark>react</mark>/",
          },
        },
        "section": Object {
          "matchLevel": "none",
          "matchedWords": Array [],
          "value": "Mocked section",
        },
        "title": Object {
          "fullyHighlighted": true,
          "matchLevel": "full",
          "matchedWords": Array [
            "react",
          ],
          "value": "<mark>React</mark>",
        },
      },
      "_snippetResult": Object {
        "body_safe": Object {
          "matchLevel": "full",
          "value": "Mocked body_safe",
        },
        "content": Object {
          "matchLevel": "full",
          "value": "MockedContent",
        },
        "excerpt": Object {
          "matchLevel": "full",
          "value": "Mocked excerpt",
        },
        "text": Object {
          "matchLevel": "full",
          "value": "Mocked text",
        },
      },
      "anchor": "mocked-anchor",
      "draft": false,
      "excerpt": "Mocked excerpt",
      "fields": Object {
        "slug": "/platforms/javascript/react/",
      },
      "objectID": "cd5528ad-935b-5a33-b25e-4f945638551b",
      "section": Object {
        "full_path": "Mocked full_path",
      },
      "title": "Mocked hit",
      "url": "/platforms/javascript/react/",
    },
  ],
  "hitsPerPage": 20,
  "index": "sentry-blog-posts",
  "nbHits": 1,
  "nbPages": 1,
  "page": 0,
  "params": "query=react&snippetEllipsisText=%E2%80%A6&highlightPreTag=%3Cmark%3E&highlightPostTag=%3C%2Fmark%3E",
  "processingTimeMS": 13,
  "query": "react",
}
`;

exports[`Search can bias individual sites 1`] = `
Array [
  Array [
    Array [
      Object {
        "indexName": "sentry-docs",
        "params": Object {
          "highlightPostTag": "</mark>",
          "highlightPreTag": "<mark>",
          "optionalFilters": Array [
            Array [
              "platforms:sentry.javascript.react",
            ],
            "legacy:0",
          ],
          "snippetEllipsisText": "…",
        },
        "query": "react",
      },
    ],
  ],
]
`;

exports[`Search does not query platforms when none are provided 1`] = `
Array [
  Array [
    Array [
      Object {
        "indexName": "sentry-docs",
        "params": Object {
          "highlightPostTag": "</mark>",
          "highlightPreTag": "<mark>",
          "optionalFilters": Array [
            "legacy:0",
          ],
          "snippetEllipsisText": "…",
        },
        "query": "react",
      },
    ],
  ],
]
`;

exports[`Search prioritizes path over platform 1`] = `
Array [
  Array [
    Array [
      Object {
        "indexName": "sentry-docs",
        "params": Object {
          "highlightPostTag": "</mark>",
          "highlightPreTag": "<mark>",
          "optionalFilters": Array [
            Array [
              "platforms:sentry.javascript.react",
            ],
            "legacy:0",
          ],
          "snippetEllipsisText": "…",
        },
        "query": "react",
      },
    ],
  ],
]
`;

exports[`Search queries 1`] = `
Array [
  Object {
    "hits": Array [
      Object {
        "context": Object {
          "context1": "Platforms > JavaScript > React",
        },
        "id": "cd5528ad-935b-5a33-b25e-4f945638551b",
        "index": "sentry-docs",
        "site": "docs",
        "text": "Mocked text",
        "title": "Mocked section",
        "url": "https://docs.sentry.io/platforms/javascript/react/",
      },
      Object {
        "context": Object {
          "context1": "Platforms > JavaScript > React",
        },
        "id": "cd5528ad-935b-5a33-b25e-4f945638551b",
        "index": "sentry-docs",
        "site": "docs",
        "text": "Mocked text",
        "title": "Mocked section",
        "url": "https://docs.sentry.io/platforms/javascript/react/",
      },
    ],
    "name": "Documentation",
    "site": "docs",
  },
]
`;

exports[`Search queries multiple indexes 1`] = `
Array [
  Object {
    "hits": Array [
      Object {
        "context": Object {
          "context1": "Platforms > JavaScript > React",
        },
        "id": "cd5528ad-935b-5a33-b25e-4f945638551b",
        "index": "sentry-docs",
        "site": "docs",
        "text": "Mocked text",
        "title": "Mocked section",
        "url": "https://docs.sentry.io/platforms/javascript/react/",
      },
      Object {
        "context": Object {
          "context1": "Platforms > JavaScript > React",
        },
        "id": "cd5528ad-935b-5a33-b25e-4f945638551b",
        "index": "sentry-docs",
        "site": "docs",
        "text": "Mocked text",
        "title": "Mocked section",
        "url": "https://docs.sentry.io/platforms/javascript/react/",
      },
    ],
    "name": "Documentation",
    "site": "docs",
  },
  Object {
    "hits": Array [
      Object {
        "context": Object {
          "context1": "Platforms > JavaScript > React",
        },
        "id": "cd5528ad-935b-5a33-b25e-4f945638551b",
        "index": "develop-docs",
        "site": "develop",
        "text": "Mocked text",
        "title": "Mocked section",
        "url": "https://develop.sentry.dev/platforms/javascript/react/",
      },
      Object {
        "context": Object {
          "context1": "Platforms > JavaScript > React",
        },
        "id": "cd5528ad-935b-5a33-b25e-4f945638551b",
        "index": "develop-docs",
        "site": "develop",
        "text": "Mocked text",
        "title": "Mocked section",
        "url": "https://develop.sentry.dev/platforms/javascript/react/",
      },
    ],
    "name": "Developer Documentation",
    "site": "develop",
  },
  Object {
    "hits": Array [
      Object {
        "context": Object {
          "context1": "Mocked hit",
        },
        "id": "cd5528ad-935b-5a33-b25e-4f945638551b",
        "index": "sentry-blog-posts",
        "site": "blog",
        "text": "Mocked text",
        "title": "Mocked section",
        "url": "https://blog.sentry.io/platforms/javascript/react/",
      },
      Object {
        "context": Object {
          "context1": "Mocked hit",
        },
        "id": "cd5528ad-935b-5a33-b25e-4f945638551b",
        "index": "sentry-blog-posts",
        "site": "blog",
        "text": "Mocked text",
        "title": "Mocked section",
        "url": "https://blog.sentry.io/platforms/javascript/react/#mocked-anchor",
      },
    ],
    "name": "Blog Posts",
    "site": "blog",
  },
  Object {
    "hits": Array [
      Object {
        "context": Object {},
        "id": "cd5528ad-935b-5a33-b25e-4f945638551b",
        "index": "sentry-help",
        "site": "blog",
        "text": "Mocked text",
        "title": "Mocked section",
        "url": "https://help.sentry.io/platforms/javascript/react/",
      },
      Object {
        "context": Object {},
        "id": "cd5528ad-935b-5a33-b25e-4f945638551b",
        "index": "sentry-help",
        "site": "blog",
        "text": "Mocked text",
        "title": "Mocked section",
        "url": "https://help.sentry.io/platforms/javascript/react/#mocked-anchor",
      },
    ],
    "name": "Help Center",
    "site": "help-center",
  },
]
`;

exports[`Search queries with path as priority 1`] = `
Array [
  Array [
    Array [
      Object {
        "indexName": "sentry-docs",
        "params": Object {
          "highlightPostTag": "</mark>",
          "highlightPreTag": "<mark>",
          "optionalFilters": Array [
            "legacy:0",
          ],
          "snippetEllipsisText": "…",
        },
        "query": "react",
      },
    ],
  ],
]
`;

exports[`Search queries with platform priority 1`] = `
Array [
  Array [
    Array [
      Object {
        "indexName": "sentry-docs",
        "params": Object {
          "highlightPostTag": "</mark>",
          "highlightPreTag": "<mark>",
          "optionalFilters": Array [
            Array [
              "platforms:sentry.javascript.react",
            ],
            "legacy:0",
          ],
          "snippetEllipsisText": "…",
        },
        "query": "react",
      },
    ],
  ],
]
`;

exports[`Search rejects unknown sites 1`] = `"SentryGlobalSearchError: unknown site \\"foobar\\" in config.include"`;

exports[`Search requires configuration 1`] = `"SentryGlobalSearchError: SentryGlobalSearch must be initialized with an array that includes one or more of: docs, develop, help-center, blog."`;

exports[`Search requires configuration as array 1`] = `"SentryGlobalSearchError: SentryGlobalSearch must be initialized with an array that includes one or more of: docs, develop, help-center, blog."`;
